editor/src/clj/util/http_client.clj

```diff
@@ -13,103 +13,56 @@
-   Note that where Ring uses InputStreams for the request and response bodies,
-   the clj-http uses ByteArrays for the bodies."
-  [{:keys [request-method scheme server-name server-port uri query-string
-           headers content-type character-encoding body socket-timeout
-           conn-timeout multipart debug insecure? save-request? follow-redirects
-           chunk-size] :as req}]
-  (let [http-url (str (name scheme) "://" server-name
-                      (when server-port (str ":" server-port))
-                      uri
-                      (when query-string (str "?" query-string)))
-        ^HttpURLConnection conn (.openConnection ^URL (URL. http-url))]
-    (when (and content-type character-encoding)
-      (.setRequestProperty conn "Content-Type" (str content-type
-                                                    "; charset="
-                                                    character-encoding)))
-    (when (and content-type (not character-encoding))
-      (.setRequestProperty conn "Content-Type" content-type))
-    (doseq [[h v] headers]
-      (.setRequestProperty conn h v))
-    (when (false? follow-redirects)
-      (.setInstanceFollowRedirects ^HttpURLConnection conn false))
-    (.setRequestMethod ^HttpURLConnection conn (.toUpperCase (name request-method)))
-    (when body
-      (.setDoOutput conn true))
-    (when socket-timeout
-      (.setReadTimeout conn socket-timeout))
-    (when conn-timeout
-      (.setConnectTimeout conn conn-timeout))
-    (when chunk-size
-      (.setChunkedStreamingMode conn chunk-size))
-    (.connect conn)
-    (when body
-      (with-open [out (.getOutputStream conn)]
-        (io/copy body out)))
-    (merge {:headers (parse-headers conn)
-            :status (.getResponseCode ^HttpURLConnection conn)
-            :body (when-not (= request-method :head)
-                    (coerce-body-entity req conn))}
-           (when save-request?
-             {:request (assoc (dissoc req :save-request?)
-                              :http-url http-url)}))))
+  Returns a CompletableFuture that will receive a map with the following keys:
+    :status     HTTP status code, e.g. 200
+    :headers    response headers map, string->(string|string[])
+    :body       the key is present only if :as was provided; either string,
+                InputStream or byte array"
+  [url & {:keys [method body headers as]
+          :or {method "GET"}}]
+  {:pre [(string? url)]}
+  (-> client
+      (.sendAsync
+        (-> (HttpRequest/newBuilder (URI. url))
+            (as-> $ (reduce-kv HttpRequest$Builder/.header $ headers))
+            (.method method (cond
+                              (nil? body) (HttpRequest$BodyPublishers/noBody)
+                              (string? body) (HttpRequest$BodyPublishers/ofString body)
+                              (instance? InputStream body) (HttpRequest$BodyPublishers/ofInputStream (fn [] body))
```

Comment:
Another Clojure 1.12 feature in action: `HttpRequest$BodyPublishers/ofInputStream` expects a Supplier instance, but we can now use fn instead of reifying a Supplier.